# RESOLVENDO PROBLEMAS DE NEGÓCIO:

QUAL OU QUAIS VARIÁVEIS ECONÔMICA INFLUENCIAM AUTOMAÇÃO INDUSTRIAL (MILHARES DE REAIS) BASE DE DADOS ABINEE?

É POSSÍVEL PREDIZER O FATURAMENTO DE AUTOMAÇÃO INDUSTRIAL POR MEIO DE MODELOS DE MACHINE LEARNING E REGRESSÃO LINEAR?

A Área de Automação Industrial inclui, equipamentos para controle e supervisão de processos, manufatura, integradores de sistemas, controladores lógicos e numéricos, instrumentos de medição de grandezas elétricas e não elétricas, sensores, equipamentos de alarme e segurança, equipamentos eletroeletrônicos para uso médico-hospitalar, entre outros.
http://www.abinee.org.br/abinee/ascor/util.htm


In [294]:
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import statsmodels.api as sm
import numpy as np
import seaborn as sns
import warnings
warnings.filterwarnings("ignore")
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import train_test_split


In [295]:
# ler a planilha de dados
df=pd.read_excel("AUT_IND_IBGE_JULHO_2023.xlsx")
df

ANO  AUTOM_IND       PIB  Indústria - total  FBCap_fixo  Agropecuária  \
0   1996        536    854763             190955      159334         40759   
1   1997        582    952089             215117      182067         44605   
2   1998        614   1002351             221709      185859         47613   
3   1999        795   1087711             236542      185088         50534   
4   2000        986   1199093             275870      219488         56963   
5   2001       1202   1315756             297881      242336         63169   
6   2002       1472   1488788             334908      266883         81515   
7   2003       1721   1717951             396568      285262        105950   
8   2004       2090   1957750             475863      339086        110912   
9   2005       2330   2170584             524687      370219        100958   
10  2006       2708   2409450             567282      414674        105294   
11  2007       3097   2720263             629072      489532        120152   
12  2008       3446   3109803             717907      602846        142052   
13  2009       2943   3333039             729222      636677        149212   
14  2010       3237   3885847             904157      797946        159932   
15  2011       3725   4376382            1011034      901928        190024   
16  2012       3920   4814760            1065681      997461        200695   
17  2013       4368   5331619            1131626     1114944        240290   
18  2014       4523   5778953            1183094     1148453        249976   
19  2015       4508   5995787            1160787     1069397        258967   
20  2016       4167   6269327            1150720      973270        306654   
21  2017       4489   6585479            1197800      958779        302972   
22  2018       5099   7004141            1313209     1057408        309612   
23  2019       5480   7389131            1385804     1143186        310714   
24  2020       5766   7609597            1484337     1260228        434621   
25  2021       7040   9012142            1993799     1614782        591085   
26  2022       8504  10079676            2300127     1794223        581343   

    Construção  Serviços  Comércio  
0        55430    515584     59820  
1        64688    577020     65779  
2        69707    612623     65936  
3        66901    654842     71595  
4        71780    698493     83631  
5        70182    759371     93401  
6        81979    853791     98241  
7        67879    968199    140092  
8        82057   1075206    164631  
9        84571   1217175    198247  
10       89102   1376715    228625  
11      105870   1570306    270799  
12      114802   1766519    322109  
13      154624   1971329    361914  
14      206926   2238750    416229  
15      233543   2519403    478242  
16      265237   2827883    548340  
17      290641   3181844    614087  
18      306946   3539665    676561  
19      296018   3735847    685707  
20      275187   3962446    699768  
21      244799   4171155    746159  
22      243280   4388329    783047  
23      248642   4660166    822590  
24      267922   4675979    825346  
25      275111   5129114    965703  
26      294040   5855005   1117748

In [296]:
#Alterando a coluna "ano" de "int" para "str"
df['ANO'] = df['ANO'].astype(str)


In [297]:
import plotly.graph_objects as go

# Escolhendo as colunas desejadas
colunas_desejadas = ['ANO', 'AUTOM_IND', 'PIB','Indústria - total', 
                     'FBCap_fixo', 'Agropecuária', 'Construção', 'Serviços',
                     'Comércio']
previsao_selecionado = df[colunas_desejadas]

# Arredondando os valores para 0 casas decimais
previsao_selecionado = previsao_selecionado.round(0)

# Criando a tabela
figura_tabela_ml = go.Figure(data=[go.Table(
    header=dict(values=previsao_selecionado.columns, fill_color='paleturquoise', align='center'),
    cells=dict(values=[previsao_selecionado[col] for col in previsao_selecionado.columns], fill_color='lavender', align='center')
)])

# Ajuste da largura das colunas
column_widths = [dict(width=60) for _ in previsao_selecionado.columns]
figura_tabela_ml.update_layout(
    autosize=False,
    width=800,
    height=300,
    title="Licenciamentos e Dados Econômicos",
    margin=dict(l=20, r=20, t=40, b=20),  # Margens
    xaxis=dict(showgrid=False, zeroline=False, showline=False, showticklabels=False),
    yaxis=dict(showgrid=False, zeroline=False, showline=False, showticklabels=False),
    paper_bgcolor="white",
    font=dict(color="black"),
)

for i, col in enumerate(previsao_selecionado.columns):
    column_widths[i]["label"] = col
figura_tabela_ml.update_layout(
    autosize=False,
    width=800,
    height=300,
    title="Licenciamentos e Dados Econômicos",
    margin=dict(l=20, r=20, t=40, b=20),  # Margens
    xaxis=dict(showgrid=False, zeroline=False, showline=False, showticklabels=False),
    yaxis=dict(showgrid=False, zeroline=False, showline=False, showticklabels=False),
    paper_bgcolor="white",
    font=dict(color="black"),
)

for i, col in enumerate(previsao_selecionado.columns):
    figura_tabela_ml.update_layout(
        {'xaxis' + str(i + 1): {'title': col}}
    )

# Exibindo a tabela
figura_tabela_ml.show()


In [102]:
df.describe()

AUTOM_IND           PIB  Indústria - total    FBCap_fixo  \
count    27.000000  2.700000e+01       2.700000e+01  2.700000e+01   
mean   3309.185185  4.053786e+06       8.553984e+05  7.189391e+05   
std    2047.702466  2.726342e+06       5.549942e+05  4.690776e+05   
min     536.000000  8.547630e+05       1.909550e+05  1.593340e+05   
25%    1596.500000  1.603370e+06       3.657380e+05  2.760725e+05   
50%    3237.000000  3.333039e+06       7.292220e+05  6.366770e+05   
75%    4498.500000  6.132557e+06       1.171940e+06  1.063402e+06   
max    8504.000000  1.007968e+07       2.300127e+06  1.794223e+06   

        Agropecuária     Construção      Serviços      Comércio  
count      27.000000      27.000000  2.700000e+01  2.700000e+01  
mean   198391.592593  171402.370370  2.426028e+06  4.297906e+05  
std    152550.515520   95978.475327  1.660882e+06  3.227723e+05  
min     40759.000000   55430.000000  5.155840e+05  5.982000e+04  
25%     91236.500000   76879.500000  9.109950e+05  1.191665e+05  
50%    149212.000000  154624.000000  1.971329e+06  3.619140e+05  
75%    280969.500000  266579.500000  3.849146e+06  6.927375e+05  
max    591085.000000  306946.000000  5.855005e+06  1.117748e+06

In [103]:
import pandas as pd
import plotly.graph_objects as go


# Remover a coluna 'ANO'
df = df.drop(columns=['ANO'])

# Criar uma matriz de correlação
correlation_matrix = df.corr()

# Exibir os valores da matriz de correlação no console
print(correlation_matrix)

# Criar um traço de heatmap
heatmap_trace = go.Heatmap(
    z=correlation_matrix.values,
    x=correlation_matrix.columns,
    y=correlation_matrix.index,
    colorscale='Viridis',
    colorbar=dict(title='Correlação')
)

# Criar o objeto de layout
heatmap_layout = go.Layout(
    title='Heatmap de Correlação',
    xaxis=dict(title=''),
    yaxis=dict(title=''),
)

# Criar o objeto de figura
heatmap_figure = go.Figure(data=[heatmap_trace], layout=heatmap_layout)

# Exibir a figura
heatmap_figure.show()


                   AUTOM_IND       PIB  Indústria - total  FBCap_fixo  \
AUTOM_IND           1.000000  0.978111           0.990435    0.977681   
PIB                 0.978111  1.000000           0.986870    0.980813   
Indústria - total   0.990435  0.986870           1.000000    0.989500   
FBCap_fixo          0.977681  0.980813           0.989500    1.000000   
Agropecuária        0.956544  0.966411           0.974975    0.950045   
Construção          0.867423  0.909775           0.885172    0.935704   
Serviços            0.964807  0.997341           0.972911    0.969438   
Comércio            0.973367  0.998001           0.981617    0.982536   

                   Agropecuária  Construção  Serviços  Comércio  
AUTOM_IND              0.956544    0.867423  0.964807  0.973367  
PIB                    0.966411    0.909775  0.997341  0.998001  
Indústria - total      0.974975    0.885172  0.972911  0.981617  
FBCap_fixo             0.950045    0.935704  0.969438  0.982536  
Agropecuária

In [123]:
import pandas as pd


# Calculando a média da coluna 'Licenciamentos'
media_ind = df['Indústria - total'].mean()

# Exibindo a média
print(f'Média do Pib Indústria: {media_ind}')


Média do Pib Indústria: 855398.4444444445


In [105]:
import pandas as pd


# Calculando a média da coluna 'Licenciamentos'
desviopadrão_ind = df['Indústria - total'].std()

# Exibindo a média
print(f'Desvio Padrão do Pib Indústria: {desviopadrão_ind}')


Desvio Padrão do Pib Indústria: 554994.2168605373


In [106]:
cv = desviopadrão_ind / media_ind
print(cv)

0.648813684973427


In [107]:
import plotly.express as px

fig_Boxplot = px.box(df, y="Indústria - total")

# Update layout to customize appearance
fig_Boxplot.update_layout(
    title="PIB DA INDÚSTRIA",
    xaxis_title="Ano",
    yaxis_title="MILHÕES DE REAIS",
    font=dict(family="Arial", size=16, color="darkblue"),  # Font style for title and axis labels
    title_font=dict(family="Arial", size=12, color="black"),  # Font style for title
    paper_bgcolor="lightgray",  # Background color of the plot
    plot_bgcolor="white",  # Background color of the plot area
)

# Update traces to customize box plot appearance
fig_Boxplot .update_traces(
    marker=dict(color="blue", line=dict(color="darkblue", width=2)),  # Box color and outline
    boxmean="sd",  # Show standard deviation as error bars
)

fig_Boxplot.show()

In [108]:
# Assume prev_ano is your DataFrame
fig_hist = px.histogram(df, x='Indústria - total')

# Update traces to customize histogram appearance
fig_hist.update_traces(
    marker_color="purple",  # Bar color
    opacity=0.7,  # Bar opacity
)

# Update layout to customize appearance
fig_hist.update_layout(
    title="PIB DA INDÚSTRIA",
    xaxis_title="MILHÕES DE REAIS",
    yaxis_title="Count",
    font=dict(family="Arial", size=16, color="darkblue"),  # Font style for title and axis labels
    title_font=dict(family="Arial", size=12, color="black"),  # Font style for title
    paper_bgcolor="lightgray",  # Background color of the plot
    plot_bgcolor="white",  # Background color of the plot area
)

# Show the plot
fig_hist.show()


In [252]:
import plotly.graph_objects as go
import numpy as np
from scipy.stats import norm

# Definindo um valor específico para calcular a probabilidade
valor_especifico = 2500000  # Substitua pelo valor desejado

# Gerando dados para a distribuição normal
x = np.linspace(media_ind - 3 * desviopadrão_ind, media_ind + 3 * desviopadrão_ind, 1000)
y = norm.cdf(x, media_ind, desviopadrão_ind)

# Calculando a probabilidade para o valor específico
probabilidade = norm.cdf(valor_especifico, media_ind, desviopadrão_ind)

# Criando o gráfico de densidade de probabilidade acumulada (CDF)
fig_cen2500 = go.Figure()

# Adicionando a curva CDF
fig_cen2500.add_trace(go.Scatter(x=x, y=y, mode='lines', name='CDF'))

# Destacando o ponto específico no gráfico
fig_cen2500.add_trace(go.Scatter(x=[valor_especifico, valor_especifico], y=[0, probabilidade],
                         mode='lines', name=f'Probabilidade: {probabilidade:.4f}',
                         line=dict(color='red', dash='dash')))

# Layout do gráfico
fig_cen2500.update_layout(title=f'Probabilidade Cumulativa (CDF) - Valor Específico: {valor_especifico}',
                  xaxis_title='PIB da Indústria (Milhões de Reais)',
                  yaxis_title='Probabilidade Cumulativa',
                  font=dict(family="Arial", size=12, color="darkblue"),  # Font style for title and axis labels
                  title_font=dict(family="Arial", size=12, color="black"),  # Font style for title
                  paper_bgcolor="lightgray",  # Background color of the plot
                  plot_bgcolor="white",  # Background color of the plot area
                  showlegend=True)

# Exibição do gráfico
fig_cen2500.show()


In [255]:
import plotly.graph_objects as go
import numpy as np
from scipy.stats import norm

# Definindo um valor específico para calcular a probabilidade
valor_especifico = 2200000  # Substitua pelo valor desejado

# Gerando dados para a distribuição normal
x = np.linspace(media_ind - 3 * desviopadrão_ind, media_ind + 3 * desviopadrão_ind, 1000)
y = norm.cdf(x, media_ind, desviopadrão_ind)

# Calculando a probabilidade para o valor específico
probabilidade = norm.cdf(valor_especifico, media_ind, desviopadrão_ind)

# Criando o gráfico de densidade de probabilidade acumulada (CDF)
fig_cen2200 = go.Figure()

# Adicionando a curva CDF
fig_cen2200.add_trace(go.Scatter(x=x, y=y, mode='lines', name='CDF'))

# Destacando o ponto específico no gráfico
fig_cen2200.add_trace(go.Scatter(x=[valor_especifico, valor_especifico], y=[0, probabilidade],
                         mode='lines', name=f'Probabilidade: {probabilidade:.4f}',
                         line=dict(color='red', dash='dash')))

# Layout do gráfico
fig_cen2200.update_layout(title=f'Probabilidade Cumulativa (CDF) - Valor Específico: {valor_especifico}',
                  xaxis_title='PIB da Indústria (Milhões de Reais)',
                  yaxis_title='Probabilidade Cumulativa',
                  font=dict(family="Arial", size=12, color="darkblue"),  # Font style for title and axis labels
                  title_font=dict(family="Arial", size=12, color="black"),  # Font style for title
                  paper_bgcolor="lightgray",  # Background color of the plot
                  plot_bgcolor="white",  # Background color of the plot area
                  showlegend=True)

# Exibição do gráfico
fig_cen2200.show()


In [256]:
import plotly.graph_objects as go
import numpy as np
from scipy.stats import norm

# Definindo um valor específico para calcular a probabilidade
valor_especifico = 2000000  # Substitua pelo valor desejado

# Gerando dados para a distribuição normal
x = np.linspace(media_ind - 3 * desviopadrão_ind, media_ind + 3 * desviopadrão_ind, 1000)
y = norm.cdf(x, media_ind, desviopadrão_ind)

# Calculando a probabilidade para o valor específico
probabilidade = norm.cdf(valor_especifico, media_ind, desviopadrão_ind)

# Criando o gráfico de densidade de probabilidade acumulada (CDF)
fig_cen2000 = go.Figure()

# Adicionando a curva CDF
fig_cen2000.add_trace(go.Scatter(x=x, y=y, mode='lines', name='CDF'))

# Destacando o ponto específico no gráfico
fig_cen2000.add_trace(go.Scatter(x=[valor_especifico, valor_especifico], y=[0, probabilidade],
                         mode='lines', name=f'Probabilidade: {probabilidade:.4f}',
                         line=dict(color='red', dash='dash')))

# Layout do gráfico
fig_cen2000.update_layout(title=f'Probabilidade Cumulativa (CDF) - Valor Específico: {valor_especifico}',
                  xaxis_title='PIB da Indústria (Milhões de Reais)',
                  yaxis_title='Probabilidade Cumulativa',
                  font=dict(family="Arial", size=12, color="darkblue"),  # Font style for title and axis labels
                  title_font=dict(family="Arial", size=12, color="black"),  # Font style for title
                  paper_bgcolor="lightgray",  # Background color of the plot
                  plot_bgcolor="white",  # Background color of the plot area
                  showlegend=True)

# Exibição do gráfico
fig_cen2000.show()


In [257]:
import plotly.graph_objects as go
import numpy as np
from scipy.stats import norm

# Definindo um valor específico para calcular a probabilidade
valor_especifico = 2300000  # Substitua pelo valor desejado

# Gerando dados para a distribuição normal
x = np.linspace(media_ind - 3 * desviopadrão_ind, media_ind + 3 * desviopadrão_ind, 1000)
y = norm.cdf(x, media_ind, desviopadrão_ind)

# Calculando a probabilidade para o valor específico
probabilidade = norm.cdf(valor_especifico, media_ind, desviopadrão_ind)

# Criando o gráfico de densidade de probabilidade acumulada (CDF)
fig_cen2300 = go.Figure()

# Adicionando a curva CDF
fig_cen2300.add_trace(go.Scatter(x=x, y=y, mode='lines', name='CDF'))

# Destacando o ponto específico no gráfico
fig_cen2300.add_trace(go.Scatter(x=[valor_especifico, valor_especifico], y=[0, probabilidade],
                         mode='lines', name=f'Probabilidade: {probabilidade:.4f}',
                         line=dict(color='red', dash='dash')))

# Layout do gráfico
fig_cen2300.update_layout(title=f'Probabilidade Cumulativa (CDF) - Valor Específico: {valor_especifico}',
                  xaxis_title='PIB da Indústria (Milhões de Reais)',
                  yaxis_title='Probabilidade Cumulativa',
                  font=dict(family="Arial", size=12, color="darkblue"),  # Font style for title and axis labels
                  title_font=dict(family="Arial", size=12, color="black"),  # Font style for title
                  paper_bgcolor="lightgray",  # Background color of the plot
                  plot_bgcolor="white",  # Background color of the plot area
                  showlegend=True)

# Exibição do gráfico
fig_cen2300.show()


In [280]:
import plotly.graph_objects as go
import numpy as np
from scipy.stats import norm

# Definindo um valor específico para calcular a probabilidade
valor_especifico = 2600000  # Substitua pelo valor desejado

# Gerando dados para a distribuição normal
x = np.linspace(media_ind - 3 * desviopadrão_ind, media_ind + 3 * desviopadrão_ind, 1000)
y = norm.cdf(x, media_ind, desviopadrão_ind)

# Calculando a probabilidade para o valor específico
probabilidade = norm.cdf(valor_especifico, media_ind, desviopadrão_ind)

# Criando o gráfico de densidade de probabilidade acumulada (CDF)
fig_cen2600 = go.Figure()

# Adicionando a curva CDF
fig_cen2600.add_trace(go.Scatter(x=x, y=y, mode='lines', name='CDF'))

# Destacando o ponto específico no gráfico
fig_cen2600.add_trace(go.Scatter(x=[valor_especifico, valor_especifico], y=[0, probabilidade],
                         mode='lines', name=f'Probabilidade: {probabilidade:.4f}',
                         line=dict(color='red', dash='dash')))

# Layout do gráfico
fig_cen2600.update_layout(title=f'Probabilidade Cumulativa (CDF) - Valor Específico: {valor_especifico}',
                  xaxis_title='PIB da Indústria (Milhões de Reais)',
                  yaxis_title='Probabilidade Cumulativa',
                  font=dict(family="Arial", size=12, color="darkblue"),  # Font style for title and axis labels
                  title_font=dict(family="Arial", size=12, color="black"),  # Font style for title
                  paper_bgcolor="lightgray",  # Background color of the plot
                  plot_bgcolor="white",  # Background color of the plot area
                  showlegend=True)

# Exibição do gráfico
fig_cen2600.show()

O coeficiente de correlação é uma medida estatística que representa a força e a direção de uma relação linear entre duas variáveis. O coeficiente de correlação varia de -1 a 1, onde -1 indica uma relação linear negativa perfeita, 0 indica nenhuma relação linear e 1 indica uma relação linear positiva perfeita. No caso, o coeficiente de correlação é 0,933.



In [116]:
import plotly.express as px

fig_Boxplot = px.box(df, y="AUTOM_IND")

# Update layout to customize appearance
fig_Boxplot.update_layout(
    title="AUTOMAÇÃO INDUSTRIAL",
    xaxis_title="Ano",
    yaxis_title="MILHÕES DE REAIS",
    font=dict(family="Arial", size=16, color="darkblue"),  # Font style for title and axis labels
    title_font=dict(family="Arial", size=12, color="black"),  # Font style for title
    paper_bgcolor="lightgray",  # Background color of the plot
    plot_bgcolor="white",  # Background color of the plot area
)

# Update traces to customize box plot appearance
fig_Boxplot .update_traces(
    marker=dict(color="green", line=dict(color="darkblue", width=2)),  # Box color and outline
    boxmean="sd",  # Show standard deviation as error bars
)

fig_Boxplot.show()

In [120]:
# Assume prev_ano is your DataFrame
fig_hist = px.histogram(df, x='AUTOM_IND')

# Update traces to customize histogram appearance
fig_hist.update_traces(
    marker_color="green",  # Bar color
    opacity=0.7,  # Bar opacity
)

# Update layout to customize appearance
fig_hist.update_layout(
    title="AUTOMAÇÃO INDUSTRIAL",
    xaxis_title="MILHÕES DE REAIS",
    yaxis_title="Count",
    font=dict(family="Arial", size=16, color="darkblue"),  # Font style for title and axis labels
    title_font=dict(family="Arial", size=12, color="black"),  # Font style for title
    paper_bgcolor="lightgray",  # Background color of the plot
    plot_bgcolor="white",  # Background color of the plot area
)

# Show the plot
fig_hist.show()


In [125]:
import pandas as pd


# Calculando a média da coluna 'Licenciamentos'
media_aut = df['AUTOM_IND'].mean()

# Exibindo a média
print(f'Média da Automação Industrial: {media_aut}')


Média da Automação Industrial: 3309.185185185185


In [126]:
import pandas as pd


# Calculando a média da coluna 'Licenciamentos'
desviopadrão_aut = df['AUTOM_IND'].std()

# Exibindo a média
print(f'Desvio Padrão do Automação Industrial: {desviopadrão_aut}')


Desvio Padrão do Automação Industrial: 2047.7024655609484


In [127]:
cv = desviopadrão_aut / media_aut
print(cv)

0.6187935552015222


In [128]:
import plotly.graph_objects as go
import numpy as np
from scipy.stats import norm

# Definindo um valor específico para calcular a probabilidade
valor_especifico = 3309  # Substitua pelo valor desejado

# Gerando dados para a distribuição normal
x = np.linspace(media_aut - 3 * desviopadrão_aut, media_aut + 3 * desviopadrão_aut, 1000)
y = norm.cdf(x, media_aut, desviopadrão_aut)

# Calculando a probabilidade para o valor específico
probabilidade = norm.cdf(valor_especifico, media_aut, desviopadrão_aut)

# Criando o gráfico de densidade de probabilidade acumulada (CDF)
fig = go.Figure()

# Adicionando a curva CDF
fig.add_trace(go.Scatter(x=x, y=y, mode='lines', name='CDF'))

# Destacando o ponto específico no gráfico
fig.add_trace(go.Scatter(x=[valor_especifico, valor_especifico], y=[0, probabilidade],
                         mode='lines', name=f'Probabilidade: {probabilidade:.4f}',
                         line=dict(color='red', dash='dash')))

# Layout do gráfico
fig.update_layout(title=f'Probabilidade Cumulativa (CDF) - Valor Específico: {valor_especifico}',
                  xaxis_title='PIB da Indústria (Milhões de Reais)',
                  yaxis_title='Probabilidade Cumulativa',
                  font=dict(family="Arial", size=12, color="darkblue"),  # Font style for title and axis labels
                  title_font=dict(family="Arial", size=12, color="black"),  # Font style for title
                  paper_bgcolor="lightgray",  # Background color of the plot
                  plot_bgcolor="white",  # Background color of the plot area
                  showlegend=True)

# Exibição do gráfico
fig.show()


In [129]:
import plotly.graph_objects as go
import numpy as np
from scipy.stats import norm

# Definindo um valor específico para calcular a probabilidade
valor_especifico = 8504  # Substitua pelo valor desejado

# Gerando dados para a distribuição normal
x = np.linspace(media_aut - 3 * desviopadrão_aut, media_aut + 3 * desviopadrão_aut, 1000)
y = norm.cdf(x, media_aut, desviopadrão_aut)

# Calculando a probabilidade para o valor específico
probabilidade = norm.cdf(valor_especifico, media_aut, desviopadrão_aut)

# Criando o gráfico de densidade de probabilidade acumulada (CDF)
fig = go.Figure()

# Adicionando a curva CDF
fig.add_trace(go.Scatter(x=x, y=y, mode='lines', name='CDF'))

# Destacando o ponto específico no gráfico
fig.add_trace(go.Scatter(x=[valor_especifico, valor_especifico], y=[0, probabilidade],
                         mode='lines', name=f'Probabilidade: {probabilidade:.4f}',
                         line=dict(color='red', dash='dash')))

# Layout do gráfico
fig.update_layout(title=f'Probabilidade Cumulativa (CDF) - Valor Específico: {valor_especifico}',
                  xaxis_title='PIB da Indústria (Milhões de Reais)',
                  yaxis_title='Probabilidade Cumulativa',
                  font=dict(family="Arial", size=12, color="darkblue"),  # Font style for title and axis labels
                  title_font=dict(family="Arial", size=12, color="black"),  # Font style for title
                  paper_bgcolor="lightgray",  # Background color of the plot
                  plot_bgcolor="white",  # Background color of the plot area
                  showlegend=True)

# Exibição do gráfico
fig.show()


In [130]:
import plotly.graph_objects as go
import numpy as np
from scipy.stats import norm

# Definindo um valor específico para calcular a probabilidade
valor_especifico = 10200  # Substitua pelo valor desejado

# Gerando dados para a distribuição normal
x = np.linspace(media_aut - 3 * desviopadrão_aut, media_aut + 3 * desviopadrão_aut, 1000)
y = norm.cdf(x, media_aut, desviopadrão_aut)

# Calculando a probabilidade para o valor específico
probabilidade = norm.cdf(valor_especifico, media_aut, desviopadrão_aut)

# Criando o gráfico de densidade de probabilidade acumulada (CDF)
fig = go.Figure()

# Adicionando a curva CDF
fig.add_trace(go.Scatter(x=x, y=y, mode='lines', name='CDF'))

# Destacando o ponto específico no gráfico
fig.add_trace(go.Scatter(x=[valor_especifico, valor_especifico], y=[0, probabilidade],
                         mode='lines', name=f'Probabilidade: {probabilidade:.4f}',
                         line=dict(color='red', dash='dash')))

# Layout do gráfico
fig.update_layout(title=f'Probabilidade Cumulativa (CDF) - Valor Específico: {valor_especifico}',
                  xaxis_title='PIB da Indústria (Milhões de Reais)',
                  yaxis_title='Probabilidade Cumulativa',
                  font=dict(family="Arial", size=12, color="darkblue"),  # Font style for title and axis labels
                  title_font=dict(family="Arial", size=12, color="black"),  # Font style for title
                  paper_bgcolor="lightgray",  # Background color of the plot
                  plot_bgcolor="white",  # Background color of the plot area
                  showlegend=True)

# Exibição do gráfico
fig.show()


In [261]:
df_1 = df[['AUTOM_IND', 'Indústria - total']]
# Exibindo o resultado
print(df_1)

    AUTOM_IND  Indústria - total
0         536             190955
1         582             215117
2         614             221709
3         795             236542
4         986             275870
5        1202             297881
6        1472             334908
7        1721             396568
8        2090             475863
9        2330             524687
10       2708             567282
11       3097             629072
12       3446             717907
13       2943             729222
14       3237             904157
15       3725            1011034
16       3920            1065681
17       4368            1131626
18       4523            1183094
19       4508            1160787
20       4167            1150720
21       4489            1197800
22       5099            1313209
23       5480            1385804
24       5766            1484337
25       7040            1993799
26       8504            2300127


In [262]:
df_1.describe().T

count           mean            std       min       25%  \
AUTOM_IND           27.0    3309.185185    2047.702466     536.0    1596.5   
Indústria - total   27.0  855398.444444  554994.216861  190955.0  365738.0   

                        50%        75%        max  
AUTOM_IND            3237.0     4498.5     8504.0  
Indústria - total  729222.0  1171940.5  2300127.0

In [263]:
# Prepara a variável de entrada X
X = np.array(df_1['Indústria - total'])
X

array([ 190955,  215117,  221709,  236542,  275870,  297881,  334908,
        396568,  475863,  524687,  567282,  629072,  717907,  729222,
        904157, 1011034, 1065681, 1131626, 1183094, 1160787, 1150720,
       1197800, 1313209, 1385804, 1484337, 1993799, 2300127], dtype=int64)

In [264]:
# Ajusta o shape de X
X = X.reshape(-1, 1)

In [265]:
# Prepara a variável alvo
y = df_1['AUTOM_IND']
y

0      536
1      582
2      614
3      795
4      986
5     1202
6     1472
7     1721
8     2090
9     2330
10    2708
11    3097
12    3446
13    2943
14    3237
15    3725
16    3920
17    4368
18    4523
19    4508
20    4167
21    4489
22    5099
23    5480
24    5766
25    7040
26    8504
Name: AUTOM_IND, dtype: int64

In [266]:
# Dividir dados em treinamento e teste
X_treino, X_teste, y_treino, y_teste = train_test_split(X, y, test_size = 0.4, random_state = 42)

In [267]:
import plotly.graph_objects as go

# Escolhendo as colunas desejadas
colunas_desejadas = ['AUTOM_IND', 'Indústria - total']
variáveis_selecionado = df_1[colunas_desejadas]

# Arredondando os valores para 0 casas decimais
variáveis_selecionado = variáveis_selecionado.round(0)

# Criando a tabela
figura_tabela_ML = go.Figure(data=[go.Table(
    header=dict(values=variáveis_selecionado.columns),
    cells=dict(values=variáveis_selecionado.transpose().values.tolist())
)])

# Layout da tabela
figura_tabela_ML.update_layout(
    title="Automação Industrial e PIB da Indústria (Milhões de Reais)",
    margin=dict(l=20, r=20, t=40, b=20),  # Margens
)

# Exibindo a tabela
figura_tabela_ML.show()



In [268]:
import plotly.express as px
import pandas as pd



# Certifique-se de que X_teste seja um DataFrame
X_teste = pd.DataFrame(X_teste)

# Adicionar uma coluna indicando se a amostra está no conjunto de treinamento ou teste
df_1['Conjunto'] = 'Treinamento'
df_1.loc[df_1.index.isin(X_teste.index), 'Conjunto'] = 'Teste'

# Criar o gráfico de dispersão
fig_traintest = px.scatter(df_1, x='Indústria - total', y='AUTOM_IND', color='Conjunto',
                           title='Divisão entre Treinamento e Teste',
                           labels={'Indústria - total': 'PIB INDÚSTRIA', 'AUTOM_IND': 'Milhões de Reais'})

# Exibir o gráfico
fig_traintest.show()


# Modelagem Preditiva (Machine Learning)
https://scikit-learn.org/stable/modules/generated/sklearn.linear_model.LinearRegression.html

In [269]:
# Cria o modelo de regressão linear simples
modelo = LinearRegression()
# Treina o modelo
modelo.fit(X_treino, y_treino)

LinearRegression()

In [270]:
y_pred=modelo.predict(X_teste)
y_pred

array([1866.73790497, 2788.87738747, 2044.440446  , 4494.33989098,
        829.77093052, 2424.36588729, 4013.47224354, 4253.48932671,
       2747.69468528, 5537.23586427,  917.71228716])

In [271]:
import plotly.graph_objects as go

# Criação do gráfico scatter
fig_retaregressão = go.Figure()

# Adição dos pontos de dispersão
fig_retaregressão.add_trace(go.Scatter(
    x=df_1['Indústria - total'],
    y=df_1['AUTOM_IND'],
    mode='markers',
    name='Dados Reais Históricos',
    marker=dict(color='blue')
))

# Adição da linha de regressão linear com cor vermelha
# Certifique-se de que X_FBCF esteja formatado corretamente para o modelo
fig_retaregressão.add_trace(go.Scatter(
    x=df_1['Indústria - total'],
    y=modelo.predict(X),  # Certifique-se de que X_FBCF está formatado corretamente
    mode='lines',
    name='Reta de Regressão com as Previsões do Modelo',
    line=dict(color='red')
))

# Layout do gráfico
fig_retaregressão.update_layout(
    title='Machine Learning Regressão Linear',
    xaxis_title='PIB DA INDÚSTRIA',
    yaxis_title='AUTOMAÇÃO INDUSTRIAL (MILHÕES DE REAIS)',
    font=dict(family="Arial", size=12, color="darkblue"),
    title_font=dict(family="Arial", size=12, color="black"),
    paper_bgcolor="lightgray",
    plot_bgcolor="white",
    showlegend=True
)

# Exibição do gráfico
fig_retaregressão.show()


In [272]:
# Avalia o modelo nos dados de teste
score = modelo.score(X_teste, y_teste)
print(f"Coeficiente R^2: {score:.2f}")

Coeficiente R^2: 0.95


In [273]:
# Intercepto - parâmetro w0
modelo.intercept_

134.76051100049744

In [274]:
# Slope - parâmetro w1
modelo.coef_

array([0.00363966])

In [275]:
import plotly.graph_objects as go
import numpy as np

# Avaliação do modelo nos dados de teste
score = modelo.score(X_teste, y_teste)
print(f"Coeficiente R^2: {score:.2f}")

# Fazendo previsões nos dados de teste
y_pred = modelo.predict(X_teste)

# Convertendo as colunas do DataFrame para arrays do NumPy
x_teste_array = np.array(X_teste)
y_teste_array = np.array(y_teste)

# Criando o gráfico scatter para mostrar os dados reais e as previsões
fig_avalmodel = go.Figure()

# Adicionando pontos de dispersão para os dados reais
fig_avalmodel.add_trace(go.Scatter(x=x_teste_array.flatten(), y=y_teste_array.flatten(), mode='markers', name='Dados Reais', marker=dict(color='blue')))

# Adicionando a linha de regressão linear para as previsões
fig_avalmodel.add_trace(go.Scatter(x=x_teste_array.flatten(), y=y_pred, mode='lines', name='Reta de Regressão (Previsões)', line=dict(color='red')))

# Layout do gráfico
fig_avalmodel.update_layout(title='Avaliação do Modelo nos Dados de Teste',
                  xaxis_title='PIB DA INDÚSTRIA (MILHÕES DE REAIS)',
                  yaxis_title='AUTOMAÇÃO INDUSTRIAL (MILHÕES DE REAIS)',
                  font=dict(family="Arial", size=12, color="darkblue"),  # Font style for title and axis labels
                  title_font=dict(family="Arial", size=12, color="black"),  # Font style for title
                  paper_bgcolor="lightgray",  # Background color of the plot
                  plot_bgcolor="white",  # Background color of the plot area
                  showlegend=True)

# Adicionando anotação para mostrar o coeficiente R^2
fig_avalmodel.add_annotation(text=f'Coeficiente R^2: {score:.2f}', xref='paper', yref='paper', x=0.95, y=0.05, showarrow=False)

# Exibição do gráfico
fig_avalmodel.show()


Coeficiente R^2: 0.95


Num modelo de regressão linear, a intercepção é o valor onde a linha de regressão cruza o eixo y quando x é zero. Também é conhecido como a intercepção y. O declive da linha de regressão é representado pelo coeficiente de x. A intercepção e o declive definem a relação linear entre duas variáveis e podem ser utilizados para estimar uma taxa média de variação.



# Implementar o Modelo
Usarei o modelo para prever o faturamento em milhões de reais para automação industrial Milhões de Reais com base no PIB DA INDÚSTRIA Milhões de Reais) Brasil do IBGE https://sidra.ibge.gov.br/tabela/1846#resultado. 

In [277]:
import plotly.graph_objects as go
import numpy as np

# Certifique-se de que 'modelo' e 'X' estão definidos corretamente no seu código
# Defina um novo valor para PIB per capita (US$ atual) do ibge
pib_ind_ibge = np.array([[2200000]])

# Faz previsão com o modelo treinado
automaçãoind_previsto = modelo.predict(pib_ind_ibge)

# Criação do gráfico scatter
fig_cen1 = go.Figure()

# Adição do ponto de dispersão para o novo valor de PIB per capita
fig_cen1.add_trace(go.Scatter(
    x=pib_ind_ibge.flatten(),
    y=automaçãoind_previsto.flatten(),
    mode='markers',
    name='Previsão para Novo Valor de PIB per capita',
    marker=dict(color='green', size=10)
))

# Adição da linha de regressão linear (ajuste para seus dados e modelo)
# Certifique-se de que X e modelo estejam definidos corretamente
fig_cen1.add_trace(go.Scatter(
    x=X.flatten(),  # Substitua isso pelos seus dados de formação de capital bruto
    y=modelo.predict(X).flatten(),  # Substitua isso pela previsão do seu modelo
    mode='lines',
    name='Reta de Regressão com as Previsões do Modelo',
    line=dict(color='red')
))

# Layout do gráfico
fig_cen1.update_layout(
    title='Previsão Cenário PIB DA INDÚSTRIA (MILHÕES DE REAIS) = 2.200.000',
    xaxis_title='MILHÕES DE REAIS',
    yaxis_title='AUTOMAÇÃO INDUSTRIAL (MILHÕES DE REAIS)',
    font=dict(family="Arial", size=12, color="darkblue"),
    title_font=dict(family="Arial", size=12, color="black"),
    paper_bgcolor="lightgray",
    plot_bgcolor="white",
    showlegend=True
)

# Exibição do gráfico
fig_cen1.show()



In [278]:
import plotly.graph_objects as go
import numpy as np

# Certifique-se de que 'modelo' e 'X' estão definidos corretamente no seu código
# Defina um novo valor para PIB per capita (US$ atual) do ibge
pib_ind_ibge = np.array([[2300000]])

# Faz previsão com o modelo treinado
automaçãoind_previsto = modelo.predict(pib_ind_ibge)

# Criação do gráfico scatter
fig_cen2 = go.Figure()

# Adição do ponto de dispersão para o novo valor de PIB per capita
fig_cen2.add_trace(go.Scatter(
    x=pib_ind_ibge.flatten(),
    y=automaçãoind_previsto.flatten(),
    mode='markers',
    name='Previsão para Novo Valor de PIB per capita',
    marker=dict(color='green', size=10)
))

# Adição da linha de regressão linear (ajuste para seus dados e modelo)
# Certifique-se de que X e modelo estejam definidos corretamente
fig_cen2.add_trace(go.Scatter(
    x=X.flatten(),  # Substitua isso pelos seus dados de formação de capital bruto
    y=modelo.predict(X).flatten(),  # Substitua isso pela previsão do seu modelo
    mode='lines',
    name='Reta de Regressão com as Previsões do Modelo',
    line=dict(color='red')
))

# Layout do gráfico
fig_cen2.update_layout(
    title='Previsão Cenário PIB DA INDÚSTRIA (MILHÕES DE REAIS) = 2.300.000',
    xaxis_title='MILHÕES DE REAIS',
    yaxis_title='AUTOMAÇÃO INDUSTRIAL (MILHÕES DE REAIS)',
    font=dict(family="Arial", size=12, color="darkblue"),
    title_font=dict(family="Arial", size=12, color="black"),
    paper_bgcolor="lightgray",
    plot_bgcolor="white",
    showlegend=True
)

# Exibição do gráfico
fig_cen2.show()



In [279]:
import plotly.graph_objects as go
import numpy as np

# Certifique-se de que 'modelo' e 'X' estão definidos corretamente no seu código
# Defina um novo valor para PIB per capita (US$ atual) do ibge
pib_ind_ibge = np.array([[2500000]])

# Faz previsão com o modelo treinado
automaçãoind_previsto = modelo.predict(pib_ind_ibge)

# Criação do gráfico scatter
fig_cen3 = go.Figure()

# Adição do ponto de dispersão para o novo valor de PIB per capita
fig_cen3.add_trace(go.Scatter(
    x=pib_ind_ibge.flatten(),
    y=automaçãoind_previsto.flatten(),
    mode='markers',
    name='Previsão para Novo Valor de PIB per capita',
    marker=dict(color='green', size=10)
))

# Adição da linha de regressão linear (ajuste para seus dados e modelo)
# Certifique-se de que X e modelo estejam definidos corretamente
fig_cen3.add_trace(go.Scatter(
    x=X.flatten(),  # Substitua isso pelos seus dados de formação de capital bruto
    y=modelo.predict(X).flatten(),  # Substitua isso pela previsão do seu modelo
    mode='lines',
    name='Reta de Regressão com as Previsões do Modelo',
    line=dict(color='red')
))

# Layout do gráfico
fig_cen3.update_layout(
    title='Previsão Cenário PIB DA INDÚSTRIA (MILHÕES DE REAIS) = 2.500.000',
    xaxis_title='MILHÕES DE REAIS',
    yaxis_title='AUTOMAÇÃO INDUSTRIAL (MILHÕES DE REAIS)',
    font=dict(family="Arial", size=12, color="darkblue"),
    title_font=dict(family="Arial", size=12, color="black"),
    paper_bgcolor="lightgray",
    plot_bgcolor="white",
    showlegend=True
)

# Exibição do gráfico
fig_cen3.show()



In [282]:
import plotly.graph_objects as go
import numpy as np

# Certifique-se de que 'modelo' e 'X' estão definidos corretamente no seu código
# Defina um novo valor para PIB per capita (US$ atual) do ibge
pib_ind_ibge = np.array([[2600000]])

# Faz previsão com o modelo treinado
automaçãoind_previsto = modelo.predict(pib_ind_ibge)

# Criação do gráfico scatter
fig_cen4 = go.Figure()

# Adição do ponto de dispersão para o novo valor de PIB per capita
fig_cen4.add_trace(go.Scatter(
    x=pib_ind_ibge.flatten(),
    y=automaçãoind_previsto.flatten(),
    mode='markers',
    name='Previsão para Novo Valor de PIB per capita',
    marker=dict(color='green', size=10)
))

# Adição da linha de regressão linear (ajuste para seus dados e modelo)
# Certifique-se de que X e modelo estejam definidos corretamente
fig_cen4.add_trace(go.Scatter(
    x=X.flatten(),  # Substitua isso pelos seus dados de formação de capital bruto
    y=modelo.predict(X).flatten(),  # Substitua isso pela previsão do seu modelo
    mode='lines',
    name='Reta de Regressão com as Previsões do Modelo',
    line=dict(color='red')
))

# Layout do gráfico
fig_cen4.update_layout(
    title='Previsão Cenário PIB DA INDÚSTRIA (MILHÕES DE REAIS) = 2.600.000',
    xaxis_title='MILHÕES DE REAIS',
    yaxis_title='AUTOMAÇÃO INDUSTRIAL (MILHÕES DE REAIS)',
    font=dict(family="Arial", size=12, color="darkblue"),
    title_font=dict(family="Arial", size=12, color="black"),
    paper_bgcolor="lightgray",
    plot_bgcolor="white",
    showlegend=True
)

# Exibição do gráfico
fig_cen4.show()



In [204]:
print(y_teste)

8     2090
13    2943
9     2330
21    4489
0      536
11    3097
16    3920
17    4368
12    3446
24    5766
1      582
Name: AUTOM_IND, dtype: int64


In [205]:
print(y_pred)

[1866.73790497 2788.87738747 2044.440446   4494.33989098  829.77093052
 2424.36588729 4013.47224354 4253.48932671 2747.69468528 5537.23586427
  917.71228716]


In [206]:
data2 = pd.read_excel('viés_bias_aut_ind.xlsx')
data2

Perído_seleção  y_teste  y_pred  Resíduos
0                8     2090    1867      -223
1               13     2943    2789      -154
2                9     2330    2044      -286
3               21     4489    4494         5
4                0      536     830       294
5               11     3097    2424      -673
6               16     3920    4013        93
7               17     4368    4253      -115
8               12     3446    2748      -698
9               24     5766    5537      -229
10               1      582     918       336

In [207]:
# Calcule o viés como a diferença média entre as previsões e os valores reais
bias = data2['y_pred'].mean() - data2['y_teste'].mean()

# Calcule a razão do viés como o viés dividido pela média dos valores reais
bias_ratio = bias / data2['y_teste'].mean()

# Exiba o viés e a razão do viés
print('Viés:', bias)
print('Razão do Viés:', bias_ratio)

Viés: -150.0
Razão do Viés: -0.04915542050227902


In [224]:
from sklearn.metrics import mean_squared_error, r2_score, mean_absolute_error, mean_absolute_percentage_error

In [225]:
# Calcular as métricas de erros
mse = mean_squared_error(y_teste, y_pred)
mae = mean_absolute_error(y_teste, y_pred)
mape = mean_absolute_percentage_error(y_teste, y_pred)
r2 = r2_score(y_teste, y_pred)

In [226]:
# Imprimir métricas de erros
print('Mean squared error: ', mse)
print('Mean absolute error: ', mae)
print('Mean absolute percentage error: ', mape)
print('R-squared: ', r2)

Mean squared error:  124402.39654655372
Mean absolute error:  282.3139863840868
Mean absolute percentage error:  0.17431022477650587
R-squared:  0.9472933505516381


# GERANDO A REGRESSÃO LINEAR MÍNIMOS QUADRADOS (ORDINARY LEAST SQUARE OLS)

A regressão por mínimos quadrados ordinários (OLS) é uma estratégia de otimização que ajuda a encontrar uma linha reta o mais próximo possível dos pontos de dados em um modelo de regressão linear. A OLS é considerada a estratégia de otimização mais útil para modelos de regressão linear, pois pode ajudá-lo a encontrar estimativas de valor real imparciais para seu alfa e beta.

In [229]:
# Definimos a variável alvo (dependente)
yI = df_1["AUTOM_IND"]
yI

0      536
1      582
2      614
3      795
4      986
5     1202
6     1472
7     1721
8     2090
9     2330
10    2708
11    3097
12    3446
13    2943
14    3237
15    3725
16    3920
17    4368
18    4523
19    4508
20    4167
21    4489
22    5099
23    5480
24    5766
25    7040
26    8504
Name: AUTOM_IND, dtype: int64

In [230]:
# Definimos a variável preditiva (independente)
XI = df_1["Indústria - total"] 
XI

0      190955
1      215117
2      221709
3      236542
4      275870
5      297881
6      334908
7      396568
8      475863
9      524687
10     567282
11     629072
12     717907
13     729222
14     904157
15    1011034
16    1065681
17    1131626
18    1183094
19    1160787
20    1150720
21    1197800
22    1313209
23    1385804
24    1484337
25    1993799
26    2300127
Name: Indústria - total, dtype: int64

In [233]:
import plotly.graph_objects as go
import statsmodels.api as sm
import pandas as pd

# Suponha que você tenha seus dados no DataFrame df
# Substitua as variáveis abaixo pelos nomes das suas colunas
XI = df_1[['Indústria - total']]  # Transforma a série em DataFrame
yI = df_1['AUTOM_IND']

# Adicione uma constante à variável independente
X = sm.add_constant(XI)

# Crie o modelo
modelo = sm.OLS(yI, X)

# Treine o modelo
resultado = modelo.fit()

# Exiba o resumo do modelo
print(resultado.summary())

# Crie um gráfico de dispersão
fig_mínimosquad = go.Figure()

fig_mínimosquad.add_trace(go.Scatter(x=XI['Indústria - total'], y=yI, mode='markers', name='Dados'))

# Adicione a linha de regressão
fig_mínimosquad.add_trace(go.Scatter(x=XI['Indústria - total'], y=resultado.predict(X),
                         mode='lines', name='Reta de Regressão'))

# Adicione o resumo do modelo como anotação no gráfico
resumo_texto = str(resultado.summary())
fig_mínimosquad.add_annotation(text=resumo_texto, xref="paper", yref="paper", x=0.5, y=0.8,
                   showarrow=False, font=dict(size=10), bgcolor="lightyellow")

# Atualize o layout do gráfico
fig_mínimosquad.update_layout(title='Regressão Linear - Resumo do Modelo Mínimos Quadrados',
                  xaxis_title='PIB DA INDÚSTRIA (MILHÕES DE REAIS)',
                  yaxis_title='AUTOMAÇÃO INDUSTRIAL (MILHÕES DE REAIS)')

# Exiba o gráfic
fig_mínimosquad.show()



                            OLS Regression Results                            
Dep. Variable:              AUTOM_IND   R-squared:                       0.981
Model:                            OLS   Adj. R-squared:                  0.980
Method:                 Least Squares   F-statistic:                     1288.
Date:                Tue, 26 Dec 2023   Prob (F-statistic):           4.99e-23
Time:                        14:11:29   Log-Likelihood:                -190.18
No. Observations:                  27   AIC:                             384.4
Df Residuals:                      25   BIC:                             387.0
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
                        coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------------
const               183.3001    103.24

 **Here's an explanation of the OLS regression results you provided:**

**Model Overview:**

- **Dependent Variable:** AUTOM_IND (presumably a measure of automation in industry)
- **Independent Variable:** Indústria - total (likely a measure of overall industry activity)

**Key Statistics:**

- **R-squared:** 0.981, indicating that 98.1% of the variation in AUTOM_IND is explained by the model.
- **Adjusted R-squared:** 0.980, a slightly adjusted R-squared accounting for the model's complexity.
- **F-statistic:** 1288, with a very low p-value (4.99e-23), suggesting strong evidence that the model is significant overall.

**Coefficients:**

- **Intercept (const):** 183.3001, representing the predicted value of AUTOM_IND when Indústria - total is zero. However, its p-value is 0.088, not quite reaching significance at the 0.05 level.
- **Indústria - total:** 0.0037, indicating that for every unit increase in Indústria - total, AUTOM_IND is estimated to increase by 0.0037 units. This coefficient has a very low p-value (0.000), suggesting strong evidence of a positive relationship.

**Diagnostics:**

- **Omnibus, Durbin-Watson, Jarque-Bera:** Tests for normality and autocorrelation of residuals. Non-significant p-values suggest no major concerns.
- **Skew, Kurtosis:** Indicate slight positive skewness and heavier tails in the residuals, but not extreme.
- **Condition Number:** High value (1.89e+06) suggests potential multicollinearity, which could affect coefficient estimates and standard errors.

**Interpretation:**

- The model suggests a strong positive relationship between Indústria - total and AUTOM_IND.
- A 1-unit increase in Indústria - total is associated with a predicted increase of 0.0037 units in AUTOM_IND.
- The model explains 98.1% of the variation in AUTOM_IND.
- Caution is advised due to potential multicollinearity, indicating a high correlation between independent variables.

**Additional Considerations:**

- **Outliers:** Check for outliers that might unduly influence the results.
- **Influential Observations:** Examine leverage and Cook's distance to identify influential points.
- **Model Assumptions:** Verify assumptions of linearity, normality of residuals, and homoscedasticity.
- **Cross-Validation:** Consider validating the model's predictive ability using a holdout sample.


In [234]:
from sklearn.linear_model import LinearRegression

In [235]:
XI = df_1.iloc[:, 1].values.reshape(-1, 1)
yI = df_1.iloc[:, 0].values.reshape(-1, 1)

In [236]:
# criar o modelo da regressão linear and e usar o método fit para os dados variável preditora XI, variável alvo yI
model = LinearRegression()
model.fit(XI, yI)

LinearRegression()

In [237]:
# Fazer previsões futuras
future_XI = np.array([[XI[-1][0] + i] for i in range(1, 2)])
future_yI_pred = model.predict(future_XI)
print(future_yI_pred)

[[8588.66469815]]


In [238]:
y_predi = model.predict(XI)
y_predi

array([[ 881.10752942],
       [ 969.4027979 ],
       [ 993.49196311],
       [1047.69623914],
       [1191.41266654],
       [1271.84752933],
       [1407.15540558],
       [1632.47972702],
       [1922.24768143],
       [2100.66536988],
       [2256.3204051 ],
       [2482.11978593],
       [2806.74979071],
       [2848.09822887],
       [3487.36371981],
       [3877.92465887],
       [4077.62135355],
       [4318.60436323],
       [4506.68402875],
       [4425.16749229],
       [4388.37962424],
       [4560.4242083 ],
       [4982.16366035],
       [5247.44778482],
       [5607.51721986],
       [7469.24572212],
       [8588.66104385]])

In [239]:
from sklearn.metrics import mean_squared_error, r2_score, mean_absolute_error, mean_absolute_percentage_error

In [240]:
# Calcular as métricas de erros
mse = mean_squared_error(yI, y_predi)
mae = mean_absolute_error(yI, y_predi)
mape = mean_absolute_percentage_error(yI, y_predi)
r2 = r2_score(yI, y_predi)

In [241]:
# Imprimir métricas de erros
print('Mean squared error: ', mse)
print('Mean absolute error: ', mae)
print('Mean absolute percentage error: ', mape)
print('R-squared: ', r2)

Mean squared error:  76870.20332194272
Mean absolute error:  222.7836337910527
Mean absolute percentage error:  0.14051197260947534
R-squared:  0.9809622885710352


In [242]:
data3 = pd.read_excel('BAIS_AUTOMAÇÃO_REGRESSÃO.xlsx')
data3

ANO  AUTOM_IND  Previsão  Resíduos
0   1996        536       881       345
1   1997        582       969       387
2   1998        614       993       379
3   1999        795      1048       253
4   2000        986      1191       205
5   2001       1202      1272        70
6   2002       1472      1407       -65
7   2003       1721      1632       -89
8   2004       2090      1922      -168
9   2005       2330      2100      -230
10  2006       2708      2256      -452
11  2007       3097      2482      -615
12  2008       3446      2807      -639
13  2009       2943      2848       -95
14  2010       3237      3487       250
15  2011       3725      3877       152
16  2012       3920      4078       158
17  2013       4368      4319       -49
18  2014       4523      4507       -16
19  2015       4508      4425       -83
20  2016       4167      4388       221
21  2017       4489      4560        71
22  2018       5099      4982      -117
23  2019       5480      5247      -233
24  2020       5766      5608      -158
25  2021       7040      7469       429
26  2022       8504      8589        85

In [243]:
# Calcular o viés da previsão
viés = data3['Previsão'] - data3['AUTOM_IND']
viés_médio = viés.mean()

# Calcular a razão do viés
média_observado = data3['AUTOM_IND'].mean()
razão_viés = viés_médio / média_observado

print(viés_médio)
print(razão_viés)

-0.14814814814814814
-4.476876930653176e-05


**Viés médio** é uma medida da diferença entre a média da distribuição da estimativa e o valor verdadeiro do parâmetro. Em outras palavras, é uma medida de quanto a estimativa está enviesada.

**Razão do viés** é a razão entre o viés médio e o valor verdadeiro do parâmetro. É uma medida da magnitude do viés.

No seu caso, o **viés médio** é de -0,14814814814814814. Isso significa que, em média, a estimativa está 0,14814814814814814 abaixo do valor verdadeiro do parâmetro.

A **razão do viés** é de -4,476876930653176e-05. Isso significa que o viés é muito pequeno em relação ao valor verdadeiro do parâmetro.

Em outras palavras, a estimativa está um pouco abaixo do valor verdadeiro do parâmetro, mas a diferença é muito pequena.

Aqui estão alguns exemplos para ilustrar o significado do viés médio e da razão do viés:

* **Viés médio = 0** significa que a estimativa é não enviesada.
* **Viés médio > 0** significa que a estimativa está tendenciosa para cima.
* **Viés médio < 0** significa que a estimativa está tendenciosa para baixo.

* **Razão do viés = 0** significa que o viés é zero.
* **Razão do viés > 0** significa que o viés é significativo.
* **Razão do viés < 0** significa que o viés é insignificante.

No seu caso, o fato de o viés médio ser pequeno e a razão do viés ser insignificante sugere que a estimativa é relativamente precisa.

In [283]:
# Criando o gráfico
fig_autprev = px.line(data3, x='ANO', y=['AUTOM_IND', 'Previsão'], title='Modelagem Regressão Mínimos Quadrados: PIB INDÚSTRIA e Previsão para AUTOMAÇÃO INDÚSTRIAL',
              labels={'value': 'Valores', 'variable': 'Variáveis'},
              line_shape='linear')

# Alterando o título do eixo x
fig_autprev.update_layout(xaxis_title='ANOS')

# Alterando a fonte e o tamanho de letra do título
fig_autprev.update_layout(title_font=dict(family='Arial', size=15))


# Exibindo o gráfico
fig_autprev.show()

In [249]:
# Criando o gráfico
fig_residuos = px.scatter(data3, x="ANO", y="Resíduos", title='Modelagem Arima: Resíduos para Automação Industrial')

# Personalizando a aparência
fig_residuos.update_traces(marker=dict(color='red', size=12),  # Altera a cor e o tamanho do marcador
                  marker_line=dict(width=2, color='blue'),  # Adiciona uma borda ao marcador
                  line=dict(color='green', width=2),  # Adiciona uma linha de ligação entre os pontos
                  selector=dict(mode='markers+lines'))  # Define o modo de exibição para marcadores e linhas

fig_residuos.update_layout(font=dict(family="Arial", size=18, color="purple"))  # Altera a fonte, tamanho e cor do texto do layout
fig_residuos.update_xaxes(title_font=dict(family="Courier", size=18, color="navy"))  # Altera a fonte, tamanho e cor do título do eixo x
fig_residuos.update_yaxes(title_font=dict(family="Courier", size=18, color="navy"))  # Altera a fonte, tamanho e cor do título do eixo y

# Exibindo o gráfico
fig_residuos.show()

In [292]:
# Importando as bibliotecas necessárias
from datetime import datetime
import dash
from dash import dcc, html
from dash.dependencies import Input, Output
import dash_bootstrap_components as dbc

# Inicializando o aplicativo Dash
app = dash.Dash(__name__, external_stylesheets=[dbc.themes.COSMO])

# Layout do aplicativo
app.layout = dbc.Container(
    [
        dbc.Tabs(
            [
                dbc.Tab(
                    label="Capa",
                    tab_id="tab-1",
                    children=[
                        dbc.Card(  # Adicionando um card como moldura
                            dbc.Row(
                                [
                                    dbc.Col(html.Img(src='https://datatem.com.br/wp-content/uploads/2020/02/automacao-industrial.png',
                                                     style={'width': '40%'}), width=6),
                                    dbc.Col(html.Img(src='https://www.predatorsoft.com/wp-content/uploads/2020/02/SOP-01-clear.png', 
                                                     style={'width': '50%'}), width=6), 
                                    
                                           

                                    dbc.Col(
                                        [
                                            html.H1("Machine Learning: Criando Cenários"),
                                            dbc.Badge('Sales & Operation Planning (S&OP)', color='success', className='ms-1',
                                                      style={'font-size': '24px'})
                                        ],
                                        width=12,
                                    ),
                                ]
                            ),
                            style={'padding': '20px'}  # Adicionando espaçamento interno ao card
                        ),

                        dbc.Row(
                            dbc.Col(
                                dbc.Card(
                                    dbc.CardBody(
                                        [
                                            html.H4("Data e Hora Atuais", className="card-title"),
                                            html.P(id="current-time", className="card-text"),
                                            dcc.Interval(
                                                id="interval-component",
                                                interval=1000,  # Atualiza a cada 1000 milissegundos (1 segundo)
                                                n_intervals=0,
                                            ),
                                        ]
                                    ),
                                    style={'padding': '20px'}  # Adicionando espaçamento interno ao card
                                )
                            )
                        )
                    ]
                ),

                dbc.Tab(
                    label="Correlações & Estatísticas",
                    tab_id="tab-2",
                    children=[
                        dbc.Card(
                            [
                                html.H1('Importância dos Dados', 
                                        style={'font-family': 'Arial, sans-serif', 'font-size': '25px', 'font-weight': 'bold'}),
                                 dcc.Markdown(
                                """
                                *Os dados são fundamentais para o sucesso do processo de Sales & Operation Planning (S&OP), 
                                que visa integrar as áreas de vendas, produção, logística e finanças de uma empresa. 
                            
                                *Os dados permitem analisar a demanda, o estoque, a capacidade produtiva, os custos 
                                e as receitas, e assim definir um plano de ação que otimize os resultados e minimize
                                os riscos. 
                            
                                *Os dados também facilitam a comunicação e a colaboração entre as diferentes áreas
                                envolvidas no S&OP, bem como o monitoramento e a avaliação do desempenho do processo.
                                """, style={'font-family': 'Arial, sans-serif', 'font-size': '16px', 'font-weight': 'bold'}
                        ),
                            ]
                        ),

                        dbc.Card(
                            [
                                html.H1('Correlações', style={'font-family': 'Arial, sans-serif', 'font-size': '25px', 'font-weight': 'bold'}),
                               ]
                        ),
                        dbc.Card(
                            [
                                dcc.Markdown(
                                        """
                                        *Correlações estatísticas são medidas que indicam o grau de associação entre
                                        duas ou mais variáveis. 
                                        
                                        *Elas podem ser positivas, negativas ou nulas, dependendo da direção e da força 
                                        da relação. 
                                        Por exemplo, se duas variáveis aumentam ou diminuem juntas, 
                                        há uma correlação positiva entre elas. 
                                        
                                        *Se uma variável aumenta e a outra diminui, há uma correlação negativa.
                                        Se não há relação entre as variações das variáveis, há uma correlação nula.
                                        
                                        """, style={'font-family': 'Arial, sans-serif', 'font-size': '16px', 'font-weight': 'bold'}),
                                
                                dbc.Row(
                                    [
                                        dbc.Col(dcc.Graph(id='figura_tabela_ml', figure=figura_tabela_ml), width=12),
                                        dbc.Col(dcc.Graph(id='heatmap-ibge', figure=heatmap_figure), width=12),
                                        html.P(),
                                    ]
                                ),
                                dbc.Row(
                                    [
                                        dbc.Col(html.H1('Estatísticas')), html.P(),
                                        dbc.Col(dcc.Graph(id='boxplot-ind', figure=fig_Boxplot), width=6), 
                                        dbc.Col(dcc.Graph(id='histograma-ind',figure=fig_hist), width=6), html.P(),
                                        
                                    ]
                                ),
                                
                                dbc.Row(
                                    [
                                        
                                        dbc.Col(dcc.Graph(id='probabilidade-ind1', figure=fig_cen2000), width=3),
                                        dbc.Col(dcc.Graph(id='probabilidade-ind2', figure=fig_cen2200), width=3),
                                        dbc.Col(dcc.Graph(id='probabilidade-ind3', figure=fig_cen2300), width=3),
                                        dbc.Col(dcc.Graph(id='probabilidade-ind4', figure=fig_cen2500), width=3),
                                                                                html.P(),
                                    ]
                                ),
                                
                                dbc.Row(
                                    [
                                        
                                        dbc.Col(dcc.Graph(id='mínimos-quadrado', figure=fig_mínimosquad), width=4),
                                        dbc.Col(dcc.Graph(id='line-regressão', figure=fig_autprev), width=4),
                                        dbc.Col(dcc.Graph(id='scatter-residuos', figure= fig_residuos), width=4),
                                        html.P(),
                                    ]
                                ),
                            ]
                        ),
                    ]
                ),

                dbc.Tab(
                    label='Modelagem Machine Learning',
                    tab_id='tab-3',
                    children=[
                        dbc.Card(
                            [
                                html.H1('Modelagem Preditiva: Machine Learning'),
                                html.P(),
                                dcc.Markdown(
                                    """
                                    *# Dividir dados em treinamento e teste X_treino, X_teste, y_treino, 
                                    y_teste = train_test_split(X, y, test_size = 0.4, random_state = 42)
                                    
                                    *O código acima é usado para dividir um conjunto de dados em dois subconjuntos: 
                                    treinamento e teste.
                                    
                                    *O conjunto de treinamento é usado para construir o modelo, enquanto o conjunto
                                    de teste é usado para avaliar a precisão do modelo. 
                                    
                                    *O train_test_split é uma função da biblioteca scikit-learn que permite dividir 
                                    um conjunto de dados em subconjuntos de treinamento e teste. 
                                    
                                    *O primeiro argumento da função é o conjunto de dados que você deseja dividir, 
                                    enquanto o segundo argumento é a variável de destino. 
                                    
                                    *O argumento test_size especifica a proporção do conjunto de dados que deve ser 
                                    usado para teste. Neste caso, 30% dos dados são usados para teste. 
                                    
                                    *O argumento random_state é usado para garantir que a divisão seja replicável.
                                    
                                    [Saiba mais sobre Linear Regression no scikit-learn]
                                    (https://scikit-learn.org/stable/modules/generated/sklearn.linear_model.LinearRegression.html)
                                    
                                    """,
                                    style={'font-family': 'Arial, sans-serif', 'font-size': '16px', 'font-weight': 'bold'}
                                ),
                                dbc.Row(
                                    [
                                        dbc.Col(dcc.Graph(id='tabela-seleção', figure=figura_tabela_ML), width=6),
                                        dbc.Col(dcc.Graph(id='scatter-treinamento-teste', figure=fig_traintest), width=6),
                                    ]
                                ),
                            ]
                        ),
                        dbc.Card(
                            id="my-card-id",  # Adicione um identificador exclusivo aqui
                            children=[
                                html.H1('Reta de Regressão Linear e Análise da Modelagem em Relação aos Dados de Teste'),
                                html.P(),
                                dbc.Row(
                                    [
                                        dbc.Col(dcc.Graph(id='scatter-retaregressão', figure=fig_retaregressão), width=6),
                                        dbc.Col(dcc.Graph(id='scatter-análisemodelo', figure=fig_avalmodel), width=6)
                                    ]
                                ),
                            ],
                            style={'padding': '20px'}  # Adicionando espaçamento interno ao card
                        ),
                    ]
                ),

                dbc.Tab(
                    label='Cenários PIB INDÚSTRIA = 2022 2.300.000 MILHÕES DE REAIS',
                    tab_id='tab-4',
                    children=[
                        dbc.Row(
                            [
                                dbc.Col(
                                    dbc.Card(
                                        [
                                            html.H1('PIB INDÚSTRIA = 2022 2.300.000 Milhões de Reais'),
                                            html.P(),
                                            dcc.Markdown(
                                                """
                                                *Uma forma de otimizar o Sales & Operation Planning é criar cenários 
                                                que simulam diferentes condições de mercado, capacidade produtiva e 
                                                rentabilidade para o negócio. Assim, é possível analisar as vantagens
                                                e desvantagens de cada cenário e escolher o mais adequado para a estratégia
                                                da empresa.
                                                
                                                *Os cenários ajudam a antecipar os riscos e as oportunidades, 
                                                bem como a definir as melhores estratégias e planos de ação. 
                                                
                                                *Além disso, os cenários facilitam a comunicação e a colaboração entre as 
                                                áreas envolvidas no processo 
                                                de planejamento integrado.
                                                
                                                *Com um PIB da Indústria em 2.300.000 Milhões de Reais,
                                                a previsão de volume é de 8.565 Milões de Reais.
                                                
                                                *No entanto, note que a probabilidade é de 99,54%, o que indica um risco 
                                                maior... isso reflete o desafio
                                                para o crescimento da linha de produtos.

                                                
                                                """,
                                                style={'font-family': 'Arial, sans-serif', 'font-size': '16px', 'font-weight': 'bold'}
                                            ),
                                        ]
                                    ),
                                    width=12
                                ),
                                dbc.Col(dcc.Graph(id='scatter-cenário2300', figure=fig_cen2), width=6),
                                dbc.Col(dcc.Graph(id='line-prob-cenário18', figure=fig_cen2300), width=6),
                            ]
                        ),
                    ]
                ),
                
                dbc.Tab(
                    label='Cenários PIB DA INDÚSTRIA = 2.200.000 MILHÕES DE REAIS 5% < 2022',
                    tab_id='tab-5',
                    children=[
                        dbc.Row(
                            [
                                dbc.Col(
                                    dbc.Card(
                                        [
                                            html.H1('PIB DA INDÚSTRIA 5% < 2022 2.200.000 Milhões de Reais'),
                                            html.P(),
                                            dcc.Markdown(
                                              """
                                              *Com um PIB da Indústria em 2.200.000 Milhões de Reais,
                                              a previsão de volume é de 8.142 Milões de Reais.
                                                
                                              *No entanto, note que a probabilidade é de 99,23%, o que indica um risco 
                                              maior... isso reflete o desafio
                                              para o crescimento da linha de produtos.

                                                """,
                                                style={'font-family': 'Arial, sans-serif', 'font-size': '16px', 'font-weight': 'bold'}
                                            ),
                                        ]
                                    ),
                                    width=12
                                ),
                                dbc.Col(dcc.Graph(id='scatter-cenário2200', figure=fig_cen1), width=6),
                                dbc.Col(dcc.Graph(id='line-prob-cenário19', figure=fig_cen2200), width=6),
                            ]
                        ),
                    ]
                ),
                
                dbc.Tab(
                    label='Cenários: PIB DA INDÚSTRIA = 2.500.000 Milhões de Reais',
                    tab_id='tab-6',
                    children=[
                        dbc.Row(
                            [
                                dbc.Col(
                                    dbc.Card(
                                        [
                                            html.H1('PIB DA INDÚSTRIA 8.6% > 2022 2.500.000 Milhões de Reais'),
                                            html.P(), 
                                            
                                             dcc.Markdown(
                                              """
                                              *Com um PIB da Indústria em 2.500.000, a previsão de receitas é de 9.234 Milhões de Reais,
                                              forte crescimento.    
                                              
                                                
                                              *No entanto, note que a probabilidade é de 99,85, o que indica um risco muito grande 
                                              somente 0,15% de chance...isso reflete o desafio
                                              para o crescimento da linha de produtos.

                                            """,
                                                style={'font-family': 'Arial, sans-serif', 'font-size': '16px', 'font-weight': 'bold'}
                                             ),
                                        ]
                                    ),
                                    width=12
                                ),
                                dbc.Col(dcc.Graph(id='scatter-cenário21', figure=fig_cen3), width=6),
                                dbc.Col(dcc.Graph(id='line-cenário21', figure=fig_cen2500), width=6),                               
                                
                               
                            ]
                        ),
                    ]
                ),
                
                dbc.Tab(
                    label='Cenários: PIB DA INDÚSTRIA = 2.600.000 Milhões de Reais',
                    tab_id='tab-7',
                    children=[
                        dbc.Row(
                            [
                                dbc.Col(
                                    dbc.Card(
                                        [
                                            html.H1('PIB DA INDÚSTRIA 13% > 2022 2.600.000 Milhões de Reais'),
                                            html.P(),  
                                            
                                            dcc.Markdown(
                                              """
                                              *Com o PIB da Indústria de 2.600.000 Milhões de Reais, 
                                              a previsão de faturamento é de 0.598 Milhões de REais, crescimento de
                                              13% em relação a 2022.
                                                
                                              *No entanto, note que a probabilidade é de 99,92%, o que indica um risco muito 
                                              alto os resultados sugerem apenas 0,08% de chance...isso reflete o desafio
                                              para o crescimento da linha de produtos.

                                            """,
                                                style={'font-family': 'Arial, sans-serif', 'font-size': '16px', 'font-weight': 'bold'}
                                             ),
                                        ]
                                    ),
                                    width=12
                                ),
                                dbc.Col(dcc.Graph(id='scatter-cenário23', figure=fig_cen4), width=6),
                                dbc.Col(dcc.Graph(id='line-prob-cenário23', figure=fig_cen2600), width=6),
                            ]
                        ),
                    ]
                ),
                
            ],
            id="tabs",
            active_tab="tab-1",  # Tab1 será a ativa inicialmente
        ),
    ],
    fluid=True
)

# Atualização da data e hora
@app.callback(
    Output("current-time", "children"),
    Input("interval-component", "n_intervals"),
)
def update_time(n):
    current_time = datetime.now().strftime("%Y-%m-%d %H:%M:%S")
    return f"Data e Hora Atuais: {current_time}"

# Executando o aplicativo Dash
if __name__ == "__main__":
    app.run_server(debug=True, port=8026)
